Introduction to service workers article added (#3159)
pwa-builder/PWABuilder: docs/home/sw-intro.md

# Introduction to Service Workers

> Service workers are on of the key components of a progressive web app, and are a requirement for your PWA to be installable and work properly offline.

Reviewer: are one* of the key components

Author: ```suggestion
Service workers are one of the key components of a progressive web app, and are a requirement for your PWA to be installable and work properly offline.
```

#### Precaching During the *Install* Event

Service workers that are being installed for the first time emit an `install` event.

We can add a listener for this event to precache essential resources for our application. This will allow our application's assets to be available when used offline.

```js
const CACHE_NAME = 'cool-cache';

// Add whichever assets you want to precache here:
const PRECACHE_ASSETS = [
    '/assets/',
    '/src/'
]

// Listener for the install event - precaches our assets list on service worker install.
self.addEventListener('install', event => {
    event.waitUntil((async () => {
        const cache = await caches.open(CACHE_NAME);
        cache.addAll(PRECACHE_ASSETS);
    })());
});
```

When the `install` event is emitted, this snippet will open a new cache with name `CACHE_NAME` and pre-cache any assets included in the `PRECACHE_ASSETS` list.

#### Claiming Clients During the *Activate* Event

After the install event, the next step in the service worker lifecycle is *activation*. An `activate` event is emitted immediately after installation is completed.

> We can use this event to have our service worker claim any clients already running:

Reviewer: What does it mean to "claim a client"

Author: ```suggestion
We can use this event to have our service worker take control of instances of our app that are already running. This is called *claiming a client*:
```

});
```

By default, a newly activated service worker won't claim any clients until they are reloaded.

Using `clients.claim()` in our activation listener tells our service worker to take control of new clients right away.

#### Defining A Fetch Strategy
Once the service worker is precaching assets, we need to provide some functionality for retrieving those assets.

We listen to the `fetch` event to allow us to intercept and handle requests for assets:

```js
self.addEventListener('fetch', event => {
  event.respondWith(async () => {
      const cache = await caches.open(CACHE_NAME);

      // match the request to our cache
      const cachedResponse = await cache.match(event.request);

      // check if we got a valid response
      if (cachedResponse !== undefined) {
          // Cache hit, return the resource
          return cachedResponse;
      } else {
        // Otherwise, go to the network
          return fetch(event.request)
      };
  });
});
```

This snippet showcases using a *Cache-First* strategy to fetch resources. When the service worker intercepts a request, it will check the cache first for a response, and then go to the network if it fails to get a response.

> The Cache-First strategy is basic and has some cons (for example, updating stale resources), but works great for simple, beginener use cases.

Reviewer: beginner*

Author: ```suggestion
The Cache-First strategy is basic and has some cons (for example, updating stale resources), but works great for simple, beginner use cases.
```
